File: Data Structures & Algorithms/buy-and-sell-crypto/submission-3.py
class Solution:
    # sliding window solution, O(n) time
    def maxProfit(self, prices: List[int]) -> int:
        result = 0
        l, r = 0, 1
        while r < len(prices):
            if prices[l] < prices[r]:
                result = max(result, prices[r] - prices[l])
            else:
                l = r
            r += 1

        return result

File: Data Structures & Algorithms/replace-elements-with-greatest-element-on-right-side/submission-3.cpp
class Solution {
public:
    // suffix max solution, O(n) time, O(1) space (does not count the output space)
    vector<int> replaceElements(vector<int>& arr) {
        int rightMax = -1;
        vector<int> result(arr.size());
        for (int i = arr.size() - 1; i >= 0; i--) {
            result[i] = rightMax;
            rightMax = max(rightMax, arr[i]);
        }
        return result;
    }
};